### ISSUE TITLE
[ENH]: Identical SVG output when running the same code twice

### ISSUE THREAD

**Issue by stefmolin** [matplotlib/matplotlib#23693]
### Problem

I have some training materials in which I use SVG plots, but every time I re-run the code the SVG that is generated changes as far as `git diff` is concerned. The SVG includes a datetime, which changes and also path IDs/xlink:href attributes. This requires more effort to make sure that nothing material has changed when inspecting the `git diff`. Is it possible to make it so that the same code will generate identical SVG output each time, just as with a PNG image?

<img width="1132" alt="Screen Shot 2022-08-20 at 9 41 53 PM" src="https://user-images.githubusercontent.com/24376333/185771779-0a165a86-67aa-47ec-a8fc-c15628be5e63.png">

**Comment by tacaswell** [comment]
The svg backend does respect `SOURCE_DATE_EPOCH`  https://github.com/matplotlib/matplotlib/blob/9b1fcf67c4228c4a2788af5bcaf0c6fde09a55bf/lib/matplotlib/backends/backend_svg.py#L407-L416

https://matplotlib.org/stable/users/prev_whats_new/whats_new_2.1.0.html?highlight=source_date_epoch#reproducible-ps-pdf-and-svg-output  has some commentary on the date and https://matplotlib.org/stable/users/prev_whats_new/whats_new_2.0.0.html?highlight=hashsalt#added-svg-hashsalt-key-to-rcparams has commentary on how to use `svg.hashsalt`.

I think the task here is to add a new tutorial (may "explain") about reproducible output with sections for the different backends.

**Comment by stefmolin** [comment]
Amazing - thanks @tacaswell! I got it working using `%config` in a notebook, which might be good to include in the tutorial as well. I'd be happy to work on the PR for this if you can provide a little guidance on what needs to be done here.

**Comment by stefmolin** [comment]
Actually, I'm having an issue using this when I generate Reveal.js slides from the notebook. I have a plot that I am building up step by step, so some of the components of the SVG have the same ID since they remain constant as the plot gets updated. This means that when I change to the slide they don't show up at all unless I refresh the page on that slide.

The SVG on the slide in question starts like this:
<img width="531" alt="Screen Shot 2022-08-21 at 1 15 57 PM" src="https://user-images.githubusercontent.com/24376333/185802978-b9f63b56-2604-4bae-810b-219d898afe66.png">

but if I refresh the page, it looks like it should:

<img width="528" alt="Screen Shot 2022-08-21 at 1 16 47 PM" src="https://user-images.githubusercontent.com/24376333/185803011-5e7f3e0b-d755-42b3-ad60-c8944d8c5001.png">

Is it possible to generate the ID for the figure as a whole and use that as the hashsalt or at least include it as part of the hashsalt set in `rcParams`? Or would I need to set a new salt in `rcParams` manually for each image?

**Comment by jklymak** [comment]
I don't understand what api you would use that would be different than manually setting the hash for each image sequence? How are you proposing we generate the hash so that it stays reproducible?

**Comment by stefmolin** [comment]
I'd love to see the ability to set the "hashsalt" (although it might need a new name) to a callable, like a random number generator on which I can set the seed. Then, each time the image is generated, a new number is generated as the salt giving the components of the SVG unique IDs. This makes it possible to set the salt just once in a notebook.

```python
# at the top of the notebook
import random
from matplotlib import rc

random.seed(42)
rc('svg', hashsalt=random.random)

# plots
...
```

**Comment by tacaswell** [comment]
For better or worse, we currently load the rcParams from a (custom) text file so putting rich Python objects (a callable) as value would be a big lift (not impossible, but it will be a lot of work on the design side and open up a much broader discussion about how we manage our global configuration (that we have been having in fits and spurts for 5 years or so)).

-------

A concern with using a function like that is that what it makes reporducible is not a single figure, but saving all of the figures in a sequence.  This makes sense for this use case, but I am not sure that it will generalize super well.   If you really do want that, I think that the management of this should be pushed up to the layer that is managing the execution / figure saving (I think jupyter in your case?).   However a major down side of this I see is that if you add or remove a figure from the sequence all subsequent svgs will change (just by virtue of the random number generator being a step farther along) 

However, I think what you want is that each cell use it's own fixed salt so out of order / rexecution ro modifications earlier or moving the cell to another notebook all are still deterministic.  It is a bit annoying, but I think that the the best path forward in the short term is probably to add 

```python
# %% cell1
mpl.rcParams ['svg.hashsart'] = 'step 1'`
....
# %% cell2
mpl.rcParams ['svg.hashsart'] = 'step 1'a`

```

to your cells. 

If you really want to make it automatic, there are a couple of ways to patch in a custom you could do something like [warning: completely untested]

```python
from itertools import count

class SaltyFigure(Figure):
    def __init__(self, *args, **kwarg):
        super().__init__(*args, **kwargs)
        self._salt_counter = count()   # or put a class s level?  can see reasons both ways.....

    def savefig(self, *args, **kwargs):
        with mpl.rc_context{{'svg.hashsalt': f'save {next(self._salt_counter)}'):
            return super().savefig(*arg, **kwargs)

fig = plt.figure(FigureClass=SaltyFigure)    # or patch into the backend
```

------


If there were to be changes to Matplotlib for this I think the right place to push would be to make the `Canvas` objects in the svg backend to carry the state for the salt rather than pulling it from rcparams on every time.  I can also see a case for pulling `uuid4` only once during the draw and doing `f'{uuid}/{N}` on each invocation of `_make_id`.  That might be valuable even if the user sets an explicit salt as now there can (per looking just at this code locally) be collisions with in a figure (if `content` is identical, but I am not sure if that is possible in SVG) which can not happen in the default case of a `uuid4` per invocation.

**Comment by stefmolin** [comment]
Thank you for coming up with a potential work around, @tacaswell. I don't want to add unnecessary extras to the slides to avoid confusing people, so I opted for something invisible to the end user: I'm wrote some JavaScript to include in my slide template that rewrites the IDs for any collisions.

I still feel like this would be nice to have in the API, as even the docs you pointed me to originally warn of collisions being an issue. If the dev team ever decides to include this in Matplotlib, I'd be happy to help work on it.

**Comment by tacaswell** [comment]
I suspect that there is something to be done here, but I am not sure what.

The things in tension are:

1. the need to ensure that we never get conflicts between figures in the same over all document (which we currently get via very random ids)
2. the need to have any given figure be fully reproducible (which you can get by setting a fixed salt)
3. doing both automatically at the correct scope

There are a couple of cases where I am not sure what the expected behavior would be

```python
fig1 = plt.figure()
fig2 = plt.figure()

some_function(fig1)
some_function(fig2)

fig1.savefig('oneA.svg')
fig1.savefig('oneB.svg')

fig2.savefig('twoA.svg')
fig2.savefig('twoB.svg')
```

It is clear that all four of these svgs should _render_ the same, however when it comes to exact byte determinism I can see a case for all being different, two sets of 2 identical svgs, and all four being the same with in the same run.   If they are not all identical, what should happen if the order of saves is permuted?

Similarly if we have 

```python
fig, ax = plt.subplots()
ax.plot(range(5))
fig.savefig('one.svg')
ln, = ax.plot(range(7))
fig.savefig('two.svg')
old_color = ln.get_color()
ln.set_color('xkcd: blue')
fig.savefig('three.svg')
ln.set_color(old_color)
fig.savefig('four.svg')
```

which of those four svg's should share any IDs?  I can see arguments for both "they should share almost all of their IDs" and "they should share none".

Given those questions I will not claim to see a path through here for the API should look like, but I think the constraints are:

 - the default behavior should stay random on every save (but exact details of the random are changeable)
 - the byte output with a fixed `svg.hashsalt` should remain unchanged (because we want to stay reproducible between mpl versions).  This is probably negotiable if you find a bug due to the current implementation (I have a worry that we could generate identical IDs within a single figure), but it would be a documented API change.
 - we have precedent of `pil_kwargs` going into `fig.savefig` so adding `svg_kwargs` is probably on the table as is adding another rcParam
 - either explain why one of the resolutions to the above questions is correct or provide enough knobs to get any of the answers
 - the API should be simpler to document and use than the proposals above that externally manage `sv.hashsalt` or post-processing the SVGs.

My knee-jerk thoughts are:

 - change the default behavior to only call uuid4 once per save and then use `f'{uuid}/{N}'` with a counter per call to `_make_id` and stay with a fixed string in the case when `svg.hashsalt` is fixed
 - add a way to inject a hashsalt per save to `savefig` and use that instead of the uuid as the base

or 

- add an rcparam for the type of salt used with the options being `{'random', 'sequential'}` where at some level we keep a count of how many times `_make_id` has been called.
- add an rcparam to control at which level (per _make_id, per save, per canvas, per process, never) the salt is updated

**Comment by stefmolin** [comment]
Personally, I would expect saving the same figure twice to result in identical files &ndash; probably also rare to need to save the same image twice and expect it to be different. For saving intermediate results (a variant of my issue above), as a user, I expect that each figure would be distinct as a whole; I can see there being benefits to being able to compare the files across each step to see the diff, but as with my case, the collisions are a concern. It seems that the desired behavior will depend on the use case, which is why I like the second option you considered above:

> * add an rcparam for the type of salt used with the options being `{'random', 'sequential'}` where at some level we keep a count of how many times `_make_id` has been called.
> * add an rcparam to control at which level (per _make_id, per save, per canvas, per process, never) the salt is updated

This gives the most control over how the IDs are created. The last part about controlling where the salt is updated is key since this gives the user the flexibility to define the desired behavior with respect to their use case, rather than deciding for them.

**Comment by tacaswell** [comment]
A better source of the "sequential" number also just occurred to me.  We have a way of notifying the `Figure` it is 'stale' when any of its children have changed.  By taking a page from the distributed computing world, if we have a "clock" on the `Figure` object that is the count of how any times that Figure has been marked stale then we have a way to differentiate saving an unchanged figure twice and mutating and then resaving the same figure (and a natural source of the `f'{N}'` in the salt).  

On a bit more consideration "per canvas" is a bad thing to expose because we can create and destroy those on behalf of the user so should not assume they can touch the canvas or store any state there.  Maybe 

- per_make_id (current default behavior)
- per_figure_mutation
- per_save
- static

is the right level to expose?  Maybe 'per_make_id' is not worth keeping and is effectively identical to per_save?

Bit annoying that we will have overlapping rcparams between `svg.hashsalt` and this new one?  Will have to work out some rules for "if hashsalt is not None, and the new one is not 'static' treat it as static and warn, if the new one is static and hashshart is `None` use the string "static" as the salt (and maybe warn?).  Maybe add an `"auto"` to be the new default which basically emulates the current behavior based on the value of `svg.hashsalt`?

-----

I think this is a cool idea and support you taking a pass at it @stefmolin , but please start knowing that this is likely to be a long process (it is adding new semantics + API and once it is released it is hard to change so we are conservative about extending things) and I suspect it will take a couple of iterations to get right.

**Comment by QuLogic** [comment]
> add a way to inject a hashsalt per save to `savefig` and use that instead of the uuid as the base

This seems the most straightforward and gives the most flexibility to the user. In fact the implementation is probably going to end up shorter than the documentation required:
```diff
diff --git a/lib/matplotlib/backends/backend_svg.py b/lib/matplotlib/backends/backend_svg.py
index 0cb6430ec8..0ac3cddad3 100644
--- a/lib/matplotlib/backends/backend_svg.py
+++ b/lib/matplotlib/backends/backend_svg.py
@@ -1,5 +1,6 @@
 import base64
 import codecs
+import contextlib
 import datetime
 import gzip
 import hashlib
@@ -1307,7 +1308,7 @@ class FigureCanvasSVG(FigureCanvasBase):
 
     fixed_dpi = 72
 
-    def print_svg(self, filename, *, bbox_inches_restore=None, metadata=None):
+    def print_svg(self, filename, *, bbox_inches_restore=None, metadata=None, hashsalt=None):
         """
         Parameters
         ----------
@@ -1338,8 +1339,15 @@ class FigureCanvasSVG(FigureCanvasBase):
             .. _DC: https://www.dublincore.org/specifications/dublin-core/
 
             __ DC_
+
+        hashsalt : str, default: :rc:`svg.hashsalt`
+            The hash salt to use when generating the SVG.
         """
-        with cbook.open_file_cm(filename, "w", encoding="utf-8") as fh:
+        with contextlib.ExitStack() as stack:
+            if hashsalt is not None:
+                stack.enter_context(mpl.rc_context({'svg.hashsalt': hashsalt}))
+            fh = stack.enter_context(
+                cbook.open_file_cm(filename, "w", encoding="utf-8"))
             if not cbook.file_requires_unicode(fh):
                 fh = codecs.getwriter('utf-8')(fh)
             dpi = self.figure.dpi
```
so that:
```python
fig1 = plt.figure()
fig2 = plt.figure()

some_function(fig1)
some_function(fig2)

fig1.savefig('oneA.svg', hashsalt='foo', metadata={'Date': None})
fig1.savefig('oneB.svg', hashsalt='foo', metadata={'Date': None})

fig2.savefig('twoA.svg')
fig2.savefig('twoB.svg')
```
would produce the same results for `one`, but different ones for `two`.
